File: ROSM3/ex01/full_name_package/src/client_name.cpp
#include "rclcpp/rclcpp.hpp"
#include "full_name_package/srv/summ_full_name.hpp"


int main(int argc, char **argv)
{
  rclcpp::init(argc, argv);

  if (argc != 4) {
    std::cerr << "Usage: client_name <last_name> <first_name> <middle_name>" << std::endl;
    return 1;
  }

  auto node = rclcpp::Node::make_shared("client_name");
  auto client = node->create_client<full_name_package::srv::SummFullName>("SummFullName");

  while (!client->wait_for_service(std::chrono::seconds(1))) {
    RCLCPP_INFO(node->get_logger(), "Waiting for service...");
  }

  auto request = std::make_shared<full_name_package::srv::SummFullName::Request>();
  request->last_name = argv[1];
  request->first_name = argv[2];
  request->middle_name = argv[3];

  auto result = client->async_send_request(request);
  if (rclcpp::spin_until_future_complete(node, result) ==
      rclcpp::FutureReturnCode::SUCCESS)
  {
    std::cout << "Результат: " << result.get()->full_name << std::endl;
  } else {
    RCLCPP_ERROR(node->get_logger(), "Failed to call service SummFullName");
  }
  rclcpp::shutdown();
  return 0;
}

File: ROSM3/ex01/ex01/ros2_ws/src/service_full_name/service_full_name/client_name.py
#!/usr/bin/env python3

import sys
import rclpy
from rclpy.node import Node
from service_full_name.srv import SummFullName

class ClientName(Node):

    def __init__(self):
        super().__init__('client_name') # вызывает конструктор родителя, задавая имя узла 'client_name'
        self.cli = self.create_client(SummFullName, 'SummFullName') # создаёт клиент для сервиса SummFullName.
        while not self.cli.wait_for_service(timeout_sec=1.0): # цикл ожидания
            self.get_logger().info('Сервис еще не доступен, ожидаем...')
        self.req = SummFullName.Request() # создаёт пустой запрос сервиса

    def send_request(self, last_name, name, first_name):
        self.req.last_name = last_name # присваивает фамилию
        self.req.name = name # имя
        self.req.first_name = first_name # отчество 
        self.future = self.cli.call_async(self.req) # вызывает сервис 

def main(args=None): 
    rclpy.init(args=args)  #  инициализирует ROS2 с аргументами

    client = ClientName() # создаёт экземпляр клиента

    if len(sys.argv) != 4:  
        client.get_logger().info('Пожалуйста, укажите 3 аргумента: фамилия имя отчество')
        sys.exit(1)

    client.send_request(sys.argv[1], sys.argv[2], sys.argv[3]) # отправляет запрос с аргументами

    while rclpy.ok(): # основной цикл ROS2 пока всё в порядке
        rclpy.spin_once(client)  # обрабатывает один цикл событий для узла
        if client.future.done():
            try:
                response = client.future.result() #  получает результат
            except Exception as e:
                client.get_logger().error(f'Ошибка вызова сервиса: {e}') # логи 
            else:
                client.get_logger().info(f'Полное имя: {response.full_name}') # выводит полное имя из ответа
            break
    # завершение 
    client.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()
